velox/connectors/hive/storage_adapters/s3fs/tests/S3InsertTest.cpp

```diff
@@ -0,0 +1,176 @@
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::connector;
+using namespace facebook::velox::connector::hive;
+using namespace facebook::velox::dwio::common;
+using namespace facebook::velox::test;
+using namespace facebook::velox::filesystems;
+
+class S3InsertTest : public testing::Test, public VectorTestBase {
+ public:
+  static constexpr char const* kMinioConnectionString{"127.0.0.1:7000"};
+  static void SetUpTestSuite() {
+    if (minioServer_ == nullptr) {
+      minioServer_ = std::make_shared<MinioServer>(kMinioConnectionString);
+      minioServer_->start();
+    }
+    ioExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(3);
+    filesystems::registerS3FileSystem();
+    auto hiveConnector =
+        connector::getConnectorFactory(
+            connector::hive::HiveConnectorFactory::kHiveConnectorName)
+            ->newConnector(
+                kHiveConnectorId,
+                minioServer_->hiveConfig(),
+                ioExecutor_.get());
+    connector::registerConnector(hiveConnector);
+  }
+
+  static void TearDownTestSuite() {
+    filesystems::finalizeS3FileSystem();
+    if (minioServer_ != nullptr) {
+      minioServer_->stop();
+      minioServer_ = nullptr;
+    }
+  }
+
+  static std::shared_ptr<MinioServer> minioServer_;
+  static std::unique_ptr<folly::IOThreadPoolExecutor> ioExecutor_;
+};
+std::shared_ptr<MinioServer> S3InsertTest::minioServer_ = nullptr;
+std::unique_ptr<folly::IOThreadPoolExecutor> S3InsertTest::ioExecutor_ =
+    nullptr;
+
+PlanNodePtr createInsertPlan(
+    PlanBuilder& inputPlan,
+    const RowTypePtr& outputRowType,
+    const std::string& outputDirectoryPath,
+    const std::vector<std::string>& partitionedBy = {},
+    const std::shared_ptr<HiveBucketProperty>& bucketProperty = {},
+    const connector::hive::LocationHandle::TableType& outputTableType =
+        connector::hive::LocationHandle::TableType::kNew,
+    const CommitStrategy& outputCommitStrategy = CommitStrategy::kNoCommit) {
+  auto insertTableHandle = std::make_shared<core::InsertTableHandle>(
+      kHiveConnectorId,
+      HiveConnectorTestBase::makeHiveInsertTableHandle(
+          outputRowType->names(),
+          outputRowType->children(),
+          partitionedBy,
+          bucketProperty,
+          HiveConnectorTestBase::makeLocationHandle(
+              outputDirectoryPath, std::nullopt, outputTableType),
+          FileFormat::PARQUET));
+
+  auto insertPlan = inputPlan.tableWrite(
+      inputPlan.planNode()->outputType(),
+      outputRowType->names(),
+      nullptr,
+      insertTableHandle,
+      bucketProperty != nullptr,
+      outputCommitStrategy);
+  return insertPlan.planNode();
+}
+
+TEST_F(S3InsertTest, s3InsertTest) {
+  const int64_t kExpectedRows = 1'000;
+  const char* kOutputDirectory = "s3://writedata/";
+
+  auto rowType = ROW(
+      {"c0", "c1", "c2", "c3"}, {BIGINT(), INTEGER(), SMALLINT(), DOUBLE()});
+
+  auto input = makeRowVector(
```

Comment:
could also use VectorFuzzer to generate some actual data, but it adds a dependency.

Reply:
The kind of input data does not matter for this test. We simply test the `write S3 -> read from S3` round trip starting from the TableWrite.
I don't think a fuzzer can potentially unearth a bug here.